# This file was generated by Rcpp::compileAttributes
# Generator token: 10BE3573-1514-4C36-9D1C-5A225CD40393

#' Calculate all burden statistics simultaneously
#' @param ccdata A matrix of markers (columns) and individuals (rows).  Data are coded as the number of copies of the minor allele.
#' @param ccstatus A vector of binary phenotype labels.  0 = control, 1 = case.
#' @param esm_K The number of markers to use in the calculation of ESM_K
#' @param LLc_maf For Li and Leal's statistic, only consider variants whose minor allele frequencies are <= maf
#' @param LLc_maf_control  For Li and Leal's statistic, calculate MAF from controls only if TRUE, otherwise from entire sample
#' @param normalize_calpha If TRUE, return T/sqrt(Z), otherwise return T.
#' @param simplecount_calpha see Details
#' @return A list of values for all burden statistics
#' @references Li, B., & Leal, S. (2008). Methods for detecting associations with rare variants for common diseases: application to analysis of sequence data. The American Journal of Human Genetics, 83(3), 311-321.
#' @references Neale, B. M., Rivas, M. A., Voight, B. F., Altshuler, D., Devlin, B., Orho-Melander, M., et al. (2011). Testing for an Unusual Distribution of Rare Variants. PLoS Genetics, 7(3), e1001322. doi:10.1371/journal.pgen.1001322
#' @references Madsen, B. E., & Browning, S. R. (2009). A groupwise association test for rare mutations using a weighted sum statistic. PLoS Genetics, 5(2), e1000384. doi:10.1371/journal.pgen.1000384
#' @references Thornton, K. R., Foran, A. J., & Long, A. D. (2013). Properties and Modeling of GWAS when Complex Disease Risk Is Due to Non-Complementing, Deleterious Mutations in Genes of Large Effect. PLoS Genetics, 9(2), e1003258. doi:10.1371/journal.pgen.1003258
#' @details  When simplecount_alpha = FALSE, heterozygous and homozygous genotypes are treated as different numbers of observations
#' of the mutation.  In other wordes, simplecounts = FALSE is equivalent to colSums( ccdata[status==1,] ).  When simplecounts=TRUE,
#' all nonzero genotype values are treated as the value 1, equivalent to  apply(data[status==1,], 2, function(x) sum(x>0, na.rm=TRUE)).
#' The latter method is used by the R package AssotesteR.
allBurdenStats <- function(ccdata, ccstatus, esm_K, LLc_maf, LLc_maf_control = TRUE, normalize_calpha = FALSE, simplecount_calpha = FALSE) {
    .Call('buRden_allBurdenStats', PACKAGE = 'buRden', ccdata, ccstatus, esm_K, LLc_maf, LLc_maf_control, normalize_calpha, simplecount_calpha)
}

#' Estimate p-values for all burden statistics by permutation
#' @param ccdata A matrix of markers (columns) and individuals (rows).  Data are coded as the number of copies of the minor allele.
#' @param ccstatus A vector of binary phenotype labels.  0 = control, 1 = case.
#' @param nperms Number of permutations to perform
#' @param esm_K The number of markers to use in the calculation of ESM_K
#' @param LLc_maf For Li and Leal's statistic, only consider variants whose minor allele frequencies are <= maf
#' @param LLc_maf_control  For Li and Leal's statistic, calculate MAF from controls only if TRUE, otherwise from entire sample
#' @param normalize_calpha If TRUE, return T/sqrt(Z), otherwise return T.
#' @param simplecount_calpha see Details
#' @return A list of p-values for all burden statistics.
#' @references Li, B., & Leal, S. (2008). Methods for detecting associations with rare variants for common diseases: application to analysis of sequence data. The American Journal of Human Genetics, 83(3), 311-321.
#' @references Neale, B. M., Rivas, M. A., Voight, B. F., Altshuler, D., Devlin, B., Orho-Melander, M., et al. (2011). Testing for an Unusual Distribution of Rare Variants. PLoS Genetics, 7(3), e1001322. doi:10.1371/journal.pgen.1001322
#' @references Madsen, B. E., & Browning, S. R. (2009). A groupwise association test for rare mutations using a weighted sum statistic. PLoS Genetics, 5(2), e1000384. doi:10.1371/journal.pgen.1000384
#' @references Thornton, K. R., Foran, A. J., & Long, A. D. (2013). Properties and Modeling of GWAS when Complex Disease Risk Is Due to Non-Complementing, Deleterious Mutations in Genes of Large Effect. PLoS Genetics, 9(2), e1003258. doi:10.1371/journal.pgen.1003258
#' @details  When simplecount_alpha = FALSE, heterozygous and homozygous genotypes are treated as different numbers of observations
#' of the mutation.  In other wordes, simplecounts = FALSE is equivalent to colSums( ccdata[status==1,] ).  When simplecounts=TRUE,
#' all nonzero genotype values are treated as the value 1, equivalent to  apply(data[status==1,], 2, function(x) sum(x>0, na.rm=TRUE)).
#' The latter method is used by the R package AssotesteR.
allBurdenStatsPerm <- function(ccdata, ccstatus, nperms, esm_K, LLc_maf, LLc_maf_control = TRUE, normalize_calpha = FALSE, simplecount_calpha = FALSE) {
    .Call('buRden_allBurdenStatsPerm', PACKAGE = 'buRden', ccdata, ccstatus, nperms, esm_K, LLc_maf, LLc_maf_control, normalize_calpha, simplecount_calpha)
}

#' The c-alpha statistic
#' @param ccdata A matrix of markers (columns) and individuals (rows).  Data are coded as the number of copies of the minor allele.
#' @param ccstatus A vector of binary phenotype labels.  0 = control, 1 = case.
#' @param normalize Return the statistic divided by the square root of its variance.
#' @param simplecounts See Details.
#' @return The c-alpha test statistic.  If normalize = TRUE, then T/sqrt(Z) is returned, otherwise T is returned.
#' @details  When simplecounts = FALSE, heterozygous and homozygous genotypes are treated as different numbers of observations
#' of the mutation.  In other wordes, simplecounts = FALSE is equivalent to colSums( ccdata[status==1,] ).  When simplecounts=TRUE,
#' all nonzero genotype values are treated as the value 1, equivalent to  apply(data[status==1,], 2, function(x) sum(x>0, na.rm=TRUE)).
#' The latter method is used by the R package AssotesteR.  
#' @references Neale, B. M., Rivas, M. A., Voight, B. F., Altshuler, D., Devlin, B., Orho-Melander, M., et al. (2011). Testing for an Unusual Distribution of Rare Variants. PLoS Genetics, 7(3), e1001322. doi:10.1371/journal.pgen.1001322
#' @examples
#' data(rec.ccdata)
#' status = c( rep(0,rec.ccdata$ncontrols),rep(1,rec.ccdata$ncases) )
#' #get minor allele freqs in the data
#' rec.ccdata.MAFS = colSums( rec.ccdata$genos[which(status==0),] )/(2*rec.ccdata$ncontrols)
#' rec.ccdata.calpha = cAlpha(rec.ccdata$genos[,which(rec.ccdata.MAFS <= 0.05)],status)
cAlpha <- function(ccdata, ccstatus, normalize = FALSE, simplecounts = FALSE) {
    .Call('buRden_cAlpha', PACKAGE = 'buRden', ccdata, ccstatus, normalize, simplecounts)
}

#' Permutation distribution of the c-alpha statistic
#' @param ccdata A matrix of markers (columns) and individuals (rows).  Data are coded as the number of copies of the minor allele.
#' @param ccstatus A vector of binary phenotype labels.  0 = control, 1 = case.
#' @param nperms The number of permutations to perform
#' @param simplecounts See Details.
#' @return The distribution of the test statistic after nperms swapping of case/control labels
#' @references Neale, B. M., Rivas, M. A., Voight, B. F., Altshuler, D., Devlin, B., Orho-Melander, M., et al. (2011). Testing for an Unusual Distribution of Rare Variants. PLoS Genetics, 7(3), e1001322. doi:10.1371/journal.pgen.1001322
#' @details  When simplecounts = FALSE, heterozygous and homozygous genotypes are treated as different numbers of observations
#' of the mutation.  In other wordes, simplecounts = FALSE is equivalent to colSums( ccdata[status==1,] ).  When simplecounts=TRUE,
#' all nonzero genotype values are treated as the value 1, equivalent to  apply(data[status==1,], 2, function(x) sum(x>0, na.rm=TRUE)).
#' The latter method is used by the R package AssotesteR.  
#' @examples
#' data(rec.ccdata)
#' status = c( rep(0,rec.ccdata$ncontrols),rep(1,rec.ccdata$ncases) )
#' #get minor allele freqs in the data
#' rec.ccdata.MAFS = colSums( rec.ccdata$genos[which(status==0),] )/(2*rec.ccdata$ncontrols)
#' rec.ccdata.calpha = cAlpha(rec.ccdata$genos[,which(rec.ccdata.MAFS <= 0.05)],status)
#' rec.ccdata.calpha.permdist = cAlpha_perm(rec.ccdata$genos[,which(rec.ccdata.MAFS <= 0.05)],status,100)
cAlpha_perm <- function(ccdata, ccstatus, nperms, simplecounts = FALSE) {
    .Call('buRden_cAlpha_perm', PACKAGE = 'buRden', ccdata, ccstatus, nperms, simplecounts)
}

#' Chi-squared statistic for a 2x2 table
#' @param a An observation
#' @param b An observation
#' @param c An observation
#' @param d An observation
#' @param yates Apply continuity correction?
#' @return The equivalent of chisq.test( matrix(c(a,b,c,d),nrow=2,byrow=T),correct=yates )$statistic
#' @details Calculated internally on a log10 scale.
chisq <- function(a, b, c, d, yates) {
    .Call('buRden_chisq', PACKAGE = 'buRden', a, b, c, d, yates)
}

#' Single-marker association test based on the chi-squared statistic
#' @param ccdata A matrix of markers (columns) and individuals (rows).  Data are coded as the number of copies of the minor allele.
#' @param ccstatus A vector of binary phenotype labels.  0 = control, 1 = case.
#' @param yates Apply continuity correction?
#' @return A vector of -log10(p-values) from a chi-squared test with one degree of freedom.  The chisq test is based on a 2x2 table of minor vs major allele counts in cases vs. controls.
#' @examples
#' data(rec.ccdata)
#' status = c(rep(0,rec.ccdata$ncontrols),rep(1,rec.ccdata$ncases))
#' #Note that the result should be very very similar to logistic regression under additive model...
#' rec.ccdata.chisq = chisq_per_marker(rec.ccdata$genos, status)
chisq_per_marker <- function(ccdata, ccstatus, yates) {
    .Call('buRden_chisq_per_marker', PACKAGE = 'buRden', ccdata, ccstatus, yates)
}

#' Association stat from Thornton, Foran, and Long (2013) PLoS Genetics
#' @param scores A vector of single-marker association test scores, on a -log10 scale
#' @param K the number of markers used to calculate ESM_K
#' @return The ESM_K test statistic value
#' @references Thornton, K. R., Foran, A. J., & Long, A. D. (2013). Properties and Modeling of GWAS when Complex Disease Risk Is Due to Non-Complementing, Deleterious Mutations in Genes of Large Effect. PLoS Genetics, 9(2), e1003258. doi:10.1371/journal.pgen.1003258
#' @examples
#' data(rec.ccdata)
#' status = c(rep(0,rec.ccdata$ncontrols),rep(1,rec.ccdata$ncases))
#' #filter out common alleles and marker pairs in high LD
#' keep = filter_sites(rec.ccdata$genos,status,0,0.05,0.8)
#' rec.ccdata.chisq = chisq_per_marker(rec.ccdata$genos[,which(keep==1)],status)
#' rec.ccdata.esm = esm( rec.ccdata.chisq, 50 )
esm <- function(scores, K) {
    .Call('buRden_esm', PACKAGE = 'buRden', scores, K)
}

#' weighted verstion of Association stat from Thornton, Foran, and Long (2013) PLoS Genetics
#' @param scores A vector of single-marker association test scores, on a -log10 scale
#' @param weights A vector of weights to use for each marker, such as dbetat(MAF,1,25)
#' @param K the number of markers used to calculate ESM_K
#' @return The ESM_K test statistic value
#' @references Thornton, K. R., Foran, A. J., & Long, A. D. (2013). Properties and Modeling of GWAS when Complex Disease Risk Is Due to Non-Complementing, Deleterious Mutations in Genes of Large Effect. PLoS Genetics, 9(2), e1003258. doi:10.1371/journal.pgen.1003258
#' @examples
#' data(rec.ccdata)
#' status = c(rep(0,rec.ccdata$ncontrols),rep(1,rec.ccdata$ncases))
#' #filter out common alleles and marker pairs in high LD
#' keep = filter_sites(rec.ccdata$genos,status,0,0.05,0.8)
#' MAF = colSums(rec.ccdata$genos[,which(keep==1)])/(2*nrow(rec.ccdata$genos))
#' b.weight = dbeta(MAF,1,25)
#' lin.weight = rep(1,length(MAF))
#' rec.ccdata.chisq = chisq_per_marker(rec.ccdata$genos[,which(keep==1)],status)
#' rec.ccdata.esm = esm( rec.ccdata.chisq,50 )
#' rec.ccdata.esm.lin =  esmw( rec.ccdata.chisq,lin.weight, 50 )
#' rec.ccdata.esm.beta = esmw( rec.ccdata.chisq,b.weight, 50 )
esmw <- function(scores, weights, K) {
    .Call('buRden_esmw', PACKAGE = 'buRden', scores, weights, K)
}

#' Association stat from Thornton, Foran, and Long (2013) PLoS Genetics
#' @param ccdata A matrix of markers (columns) and individuals (rows).  Data are coded as the number of copies of the minor allele.
#' @param ccstatus A vector of binary phenotype labels.  0 = control, 1 = case.
#' @param k The number of markers for the ESM_K statistic.
#' @param yates Logical -- apply Yates' correction to chi-squared statistic?
#' @return The ESM_K test statistic value based on chi-squared tests per marker.
#' @references Thornton, K. R., Foran, A. J., & Long, A. D. (2013). Properties and Modeling of GWAS when Complex Disease Risk Is Due to Non-Complementing, Deleterious Mutations in Genes of Large Effect. PLoS Genetics, 9(2), e1003258. doi:10.1371/journal.pgen.1003258
#' @examples
#' data(rec.ccdata)
#' status = c(rep(0,rec.ccdata$ncontrols),rep(1,rec.ccdata$ncases))
#' #filter out common alleles and marker pairs in high LD
#' keep = filter_sites(rec.ccdata$genos,status,0,0.05,0.8)
#' rec.ccdata.chisq = esm_chisq( rec.ccdata$genos[,which(keep==1)], status, 50 )
esm_chisq <- function(ccdata, ccstatus, k, yates = TRUE) {
    .Call('buRden_esm_chisq', PACKAGE = 'buRden', ccdata, ccstatus, k, yates)
}

#' Obtain permutaion distribution of the ESM_K statistic for case/control data
#' @param ccdata A matrix of markers (columns) and individuals (rows).  Data are coded as the number of copies of the minor allele.
#' @param ccstatus A vector of binary phenotype labels.  0 = control, 1 = case.
#' @param nperms Number of permutations to perform
#' @param k Number of markers to use for ESM_K statistic
#' @param yates Logical -- apply Yates' correction to chi-squared statistic?
#' @return A vector of the permuted test statistic values
#' @references Thornton, K. R., Foran, A. J., & Long, A. D. (2013). Properties and Modeling of GWAS when Complex Disease Risk Is Due to Non-Complementing, Deleterious Mutations in Genes of Large Effect. PLoS Genetics, 9(2), e1003258. doi:10.1371/journal.pgen.1003258
#' @examples
#' data(rec.ccdata)
#' status = c(rep(0,rec.ccdata$ncontrols),rep(1,rec.ccdata$ncases))
#' #filter out common alleles and marker pairs in high LD
#' keep = filter_sites(rec.ccdata$genos,status,0,0.05,0.8)
#' rec.ccdata.chisq = chisq_per_marker(rec.ccdata$genos[,which(keep==1)],status)
#' rec.ccdata.esm = esm( rec.ccdata.chisq, 50 )
#' rec.ccdata.esm.permdist = esm_perm_binary(rec.ccdata$genos[,which(keep==1)],status,100,50)
esm_perm_binary <- function(ccdata, ccstatus, nperms, k, yates = TRUE) {
    .Call('buRden_esm_perm_binary', PACKAGE = 'buRden', ccdata, ccstatus, nperms, k, yates)
}

#' Apply frequency and LD filters to a genotype matrix
#' @param ccdata A matrix of markers (columns) and individuals (rows).  Data are coded as the number of copies of the minor allele.
#' @param ccstatus A vector of binary phenotype labels.  0 = control, 1 = case.
#' @param minfreq A site with minor allele frequency < minfreq will not be kept.
#' @param maxfreq A site with minor allele frequency >= maxfreq will not be kept.
#' @param rsq_cutoff  When comparing two sites, if the genotype correlation coefficient r^2 is >= rsq_cutoff, only the first site will be kept.
#' @return A vector of integers containing the values 0 (not kept) and 1 (kept).  The length of the vector is equal to the number of columns in ccdata.
#' @details Regarding rsq_cutoff, when sites i and j are compared (j > i), site i will be kept and site j will not be kept.
#' @examples
#' data(rec.ccdata)
#' status=c(rep(0,rec.ccdata$ncontrols),rep(1,rec.ccdata$ncases))
#' keep=filter_sites(rec.ccdata$genos,status,0,0.05,0.8)
filter_sites <- function(ccdata, ccstatus, minfreq, maxfreq, rsq_cutoff) {
    .Call('buRden_filter_sites', PACKAGE = 'buRden', ccdata, ccstatus, minfreq, maxfreq, rsq_cutoff)
}

#' Calculate Madsen-Browning weights.
#' @param ccdata A matrix of markers (columns) and individuals (rows).  Data are coded as the number of copies of the minor allele.
#' @param ccstatus A vector of binary phenotype labels.  0 = control, 1 = case.
#' @return An array of weights, one for each column in data.
#' @details Calculation is done under the "general genetic model" defined in Madsen and Browning.
#' @references Madsen, B. E., & Browning, S. R. (2009). A groupwise association test for rare mutations using a weighted sum statistic. PLoS Genetics, 5(2), e1000384. doi:10.1371/journal.pgen.1000384
MBweights <- function(ccdata, ccstatus) {
    .Call('buRden_MBweights', PACKAGE = 'buRden', ccdata, ccstatus)
}

#' Madsen-Browning test statistics
#' @param ccdata A matrix of markers (columns) and individuals (rows).  Data are coded as the number of copies of the minor allele.
#' @param ccstatus A vector of binary phenotype labels.  0 = control, 1 = case.
#' @return The M-B test statistic for the "general genetic", "recessive", and "dominant" models.
#' @references Madsen, B. E., & Browning, S. R. (2009). A groupwise association test for rare mutations using a weighted sum statistic. PLoS Genetics, 5(2), e1000384. doi:10.1371/journal.pgen.1000384
#' @details When calculating the rank of an individual's score, the function uses the equivalent of ties="min" in R's rank() function.
#' @examples
#' data(rec.ccdata)
#' status = c(rep(0,rec.ccdata$ncontrols),rep(1,rec.ccdata$ncases))
#' #filter out common alleles and marker pairs in high LD
#' keep = filter_sites(rec.ccdata$genos,status,0,0.05,0.8)
#' mbstats = MBstat( rec.ccdata$genos[,which(keep==1)], status )
MBstat <- function(ccdata, ccstatus) {
    .Call('buRden_MBstat', PACKAGE = 'buRden', ccdata, ccstatus)
}

#' Get permutation distribution of Madsen-Browning test statistics
#' @param ccdata A matrix of markers (columns) and individuals (rows).  Data are coded as the number of copies of the minor allele.                                                                                    
#' @param ccstatus A vector of binary phenotype labels.  0 = control, 1 = case. 
#' @param nperms The number of permutations to perform
#' @return A data frame of permuted statistics
#' @references Madsen, B. E., & Browning, S. R. (2009). A groupwise association test for rare mutations using a weighted sum statistic. PLoS Genetics, 5(2), e1000384. doi:10.1371/journal.pgen.1000384
#' @examples
#' data(rec.ccdata)
#' status = c(rep(0,rec.ccdata$ncontrols),rep(1,rec.ccdata$ncases))
#' #filter out common alleles and marker pairs in high LD
#' keep = filter_sites(rec.ccdata$genos,status,0,0.05,0.8)
#' mbstats = MBstat( rec.ccdata$genos[,which(keep==1)], status )
#' mbstats.perm = MB_perm( rec.ccdata$genos[,which(keep==1)], status, 100 )
MB_perm <- function(ccdata, ccstatus, nperms) {
    .Call('buRden_MB_perm', PACKAGE = 'buRden', ccdata, ccstatus, nperms)
}

#' Pearson's product-moment correlation
#' @param x A vector of values.
#' @param y A vector of values.
#' @return The correlation coefficient between x and y. This should/will be equal to R's cor(x,y).
#' @note This implementation is based on a copy of [libsequence's](http://github.com/molpopgen/libsequence) template function object Sequence::ProductMoment in <Sequence/Correlations.hpp>
#' @examples
#' x=rnorm(100)
#' y=rnorm(100)
#' xy.pm = ProductMoment(x,y)
#' xy.cor = cor(x,y)
#' print( xy.pm - xy.cor )
ProductMoment <- function(x, y) {
    .Call('buRden_ProductMoment', PACKAGE = 'buRden', x, y)
}

#' Calculates Li and Leal's collapsed variant statistic, v_c
#' @param ccdata A matrix of markers (columns) and individuals (rows).  Data are coded as the number of copies of the minor allele.
#' @param ccstatus A vector of binary phenotype labels.  0 = control, 1 = case.
#' @param maf Only consider variants whose minor allele frequencies are <= maf
#' @param maf_controls  If true, calculate mafs from controls only.  Otherwise, use all individuals
#' @return A chi-squared statistic based on a 2x2 table of the number of cases and controls with and without rare alleles. Yate's continuity correction is applied.
#' @references Li, B., & Leal, S. (2008). Methods for detecting associations with rare variants for common diseases: application to analysis of sequence data. The American Journal of Human Genetics, 83(3), 311-321.
#' @examples
#' data(rec.ccdata)
#' status = c(rep(0,rec.ccdata$ncontrols),rep(1,rec.ccdata$ncases))
#' LL = LLcollapse(rec.ccdata$genos,status,0.01)
LLcollapse <- function(ccdata, ccstatus, maf, maf_controls = FALSE) {
    .Call('buRden_LLcollapse', PACKAGE = 'buRden', ccdata, ccstatus, maf, maf_controls)
}

#' Permutation distribution of Li and Leal's collapsed variant statistic, v_c
#' @param ccdata A matrix of markers (columns) and individuals (rows).  Data are coded as the number of copies of the minor allele.
#' @param ccstatus A vector of binary phenotype labels.  0 = control, 1 = case.
#' @param nperms The number of permutations to perform
#' @param maf Only consider variants whose minor allele frequencies are <= maf
#' @param maf_controls  If true, calculate mafs from controls only.  Otherwise, use all individuals
#' @return The non-centrality parameter of a chi-squared distribution.  This is obtained using the proportion of controls and cases with rare variants.
#' @references Li, B., & Leal, S. (2008). Methods for detecting associations with rare variants for common diseases: application to analysis of sequence data. The American Journal of Human Genetics, 83(3), 311-321.
#' @examples
#' data(rec.ccdata)
#' status = c(rep(0,rec.ccdata$ncontrols),rep(1,rec.ccdata$ncases))
#' LL.perm = LLcollapse_perm(rec.ccdata$genos,status,10,0.01)
LLcollapse_perm <- function(ccdata, ccstatus, nperms, maf, maf_controls = FALSE) {
    .Call('buRden_LLcollapse_perm', PACKAGE = 'buRden', ccdata, ccstatus, nperms, maf, maf_controls)
}

